# RNAInfuser experiment design

## Read profile

* Read-length distributions
  * Mapped parts of the reads
  * Whole read
* Error distributions (from mapping)
  * Substitions
  * Indels
  * Noise

## RNA

* Poly-A size distributions
* Expression
  * TPM
  * R^2 (Coefficient of determination) comparison
  * PCR (make use of UMIs)
* cDNA vs direct RNA comparison

## PCR

* A histogram of number of reads per molecule (Simulated)
* A histogram of number of reads per UMI  (Both)

## Chimeric read characteristics

* Readthrough
* Fusion
* Random pairing
  * Percentage in real and simulated data

## Single cell characteristics (Not it paper?)

* Long Single Cell reads from short single cell reads
* Novel Single Cell dataset from matrix
* Random matrix

## Other papers to look at

* [Trans-NanoSim](https://academic.oup.com/gigascience/article/9/6/giaa061/5855462)
* [Minnow](https://academic.oup.com/bioinformatics/article/35/14/i136/5529127)
* [Badread](https://joss.theoj.org/papers/10.21105/joss.01316)
* [PBSIM2](https://academic.oup.com/bioinformatics/article/37/5/589/5911629)

